design/ceph/cephx-keyring-rotation.md

```diff
@@ -0,0 +1,483 @@
+
+e.g., Admin key(s), mon, mgr, osd, mds, rgw, rbdmirror's internal (non-peer) key
+
+### "Non-daemon" keys
+Any key whose rotation may reasonably require user action beyond Rook API controls.
+
+Because these keys affect non-daemon connections, Rook must make sure the user can initiate rotation
+during a maintenance window. It is imperative that the user can determine when rotation is finished,
+so they know when to update the non-daemon application with the updated key.
+
+e.g., CephClient keys, CephRBDMirror peer keys, CSI keys
+
+Because different types of key rotations may call for different administrator/user workflows, these
+rotations should be able to be initiated independently. For example, a user may have an automated
+workflow to update CephClient A's consumer, but CephClient B's consumer may be a manual process.
+CephClient A and B should allow rotation separately.
+
+CSI keys are a special case that best fits into the non-daemon case. Rotation of CSI keys will affect
+Ceph-CSI's RBD and CephFS volume mounts associated with application pod PVCs. This rotation may
+require administrators to reboot or drain/undrain all nodes with Ceph-CSI PVCs.
+
+### Overlapping rotation
+
+When CSI keys are updated, sers must drain/undrain (or reboot) all nodes to transition PVC mounts
+from using old key to new key.
+
+Users with many nodes may not be able to update them all within the (2 * `auth_service_ticket_ttl`)
+window. Some users with very large k8s clusters might only drain/reboot a portion of their nodes
+during a maintenance window over a period of several days. (Service ticket TTL could be extended,
+but extending it to several days would also have Ceph security implications that are best avoided
+if possible)
+
+In cases where the rotated key can't be picked up quickly, the alternative is for Rook to create a
+new ceph client (user) with the new key without deleting the old user/key. Then, Rook updates the
+CSI secrets to have the new user+key. This would begin the window in which users can drain/reboot
+nodes. Any new PVC mounts would use the new key, and old mounts would continue operating happily
+since the old user+key are not yet destroyed.
+
+After the user is done with node updates, the user should have some way of indicating to Rook that
+the old key is no longer needed, telling Rook it is safe to delete the old key. Autodetection might
+be possible, but let's leave that for a future exercise.
+
+For clarity, this design could be called "overlapping" rotation, and the design elsewhere in this
+doc "in-place" rotation.
+
+Overlapping rotation will be implemented for 3 key types:
+
+- CSI keys
+- RBD mirror peer keys (not daemon keys)
+- CephFS mirror peer (not daemon keys)
+
+## User interfaces
+
+### Key rotation base interface
+
+```yaml
+keyRotationPolicy: Disabled | WithCephVersionUpgrade | KeyGeneration
+keyGeneration: <int> # used with KeyGeneration
+keepPriorKeyCount: <int> # only present to configure overlapping rotation
```

Comment:
`KeyPriorCount` to make it consistent

Reply:
I still prefer this wording but am interested in what others have to say

Reply:
Interesting. That would allow us to distinguish the "unset" zero value, which we would assume means 1 (current key). I'm not sure that is more clear though, like you said. And I don't know if determining "unset" is needed or interesting internally.